docs/user-defined-type-functions.md

```diff
@@ -0,0 +1,241 @@
+| `isunion(arg: typelib)` | `boolean` | returns true if the argument is syntactically a runtime representation of the union type |
+| `isintersection(arg: typelib)` | `boolean` | returns true if the argument is syntactically a runtime representation of the intersection type |
+| `istable(arg: typelib)` | `boolean` | returns true if the argument is syntactically a runtime representation of a `table` type |
+| `ismetatable(arg: typelib)` | `boolean` | returns true if the argument is syntactically a runtime representation of `Metatable` |
+| `isfunction(arg: typelib)` | `boolean` | returns true if the argument is syntactically a runtime representation of a `function` type |
+| `isclass(arg: typelib)` | `boolean` | returns true if the argument is syntactically a runtime representation of a `class` type |
+
+#### Any
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `issubtypeof(arg: typelib)` | `boolean` | returns true if self is a subtype or equal to arg in the type hierarchy |
+| `conformsto(arg: typelib)` | `boolean` | returns true if self is equal to arg in the type hierarchy |
+
+#### Negation
+
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `gettype()` | `typelib` | returns the runtime representation of the self's type being negated |
+
+#### String
+
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `getmetatable()` | `typelib` | returns the runtime representation of self's `Metatable` |
+
+#### StringSingleton
+
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `getvalue()` | `string` | returns self's value of a string singleton |
+
+#### BooleanSingleton
+
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `getvalue()` | `boolean` | returns self's boolean singleton value of either `true` or `false` |
+
+#### Table
+
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `addprops(key: typelib, value: typelib)` | `nil` | adds a key, value pair to self's table properties; if the same key exists already, overrides the value |
+| `delprops(key: typelib)` | `nil` | removes the key from self's table properties along with the value associated with it; if the key doesn't exist, nothing happens |
+| `getprops()` | `{typelib: typelib}` | returns a table of self's table properties (e.g. `{["age"] = 20}` will return `{typelib.getstringsingleton("age") = typelib.getnumber()}`) |
+| `settableindexer(indexty: typelib, indexresultty: typelib)` | `nil` | sets self's index type to the first argument and index result type to the second |
+| `getindextype()` | `typelib` | returns self's index type |
+| `getindexresulttype()` | `typelib` | returns self's index result type |
```

Comment:
I've regretted the verbose names I've given to the `TableIndexer` fields. I wanted to get around to naming them `keyType` and `valueType`.

I want to avoid making this mistake in userspace.